server/aws-lsp-codewhisperer/src/language-server/telemetryService.ts

```diff
@@ -0,0 +1,122 @@
+    UserContext,
+    OptOutPreference,
+    SendTelemetryEventRequest,
+} from '../client/token/codewhispererbearertokenclient'
+import { getCompletionType, getLoginTypeFromProvider, LoginType } from './utils'
+import { getRuntimeLanguage } from './languageDetection'
+
+export class TelemetryService extends CodeWhispererServiceToken {
+    private userContext: UserContext | undefined
+    private optOutPreference!: OptOutPreference
+    private enableTelemetryEventsToDestination!: boolean
+    private telemetry: Telemetry
+    private loginType: LoginType
+    private credentialsType: CredentialsType
+
+    constructor(
+        credentialsProvider: CredentialsProvider,
+        credentialsType: CredentialsType,
+        telemetry: Telemetry,
+        additionalAwsConfig: AWS.ConfigurationOptions = {}
+    ) {
+        super(credentialsProvider, additionalAwsConfig)
+        this.credentialsType = credentialsType
+        this.telemetry = telemetry
+        this.loginType = getLoginTypeFromProvider(credentialsProvider)
+    }
+
+    public updateUserContext(userContext: UserContext | undefined): void {
+        this.userContext = userContext
+    }
+
+    public updateOptOutPreference(optOutPreference: OptOutPreference): void {
+        this.optOutPreference = optOutPreference
+    }
+
+    public updateEnableTelemetryEventsToDestination(enableTelemetryEventsToDestination: boolean): void {
+        this.enableTelemetryEventsToDestination = enableTelemetryEventsToDestination
+    }
+
+    private getSuggestionState(session: CodeWhispererSession): SuggestionState {
+        let suggestionState: SuggestionState
+        switch (session.getAggregatedUserTriggerDecision()) {
+            case 'Accept':
+                suggestionState = 'ACCEPT'
+                break
+            case 'Reject':
+                suggestionState = 'REJECT'
+                break
+            case 'Discard':
+                suggestionState = 'DISCARD'
+                break
+            default:
+                suggestionState = 'EMPTY'
+        }
+        return suggestionState
+    }
+
+    private isLoginInvalidForTelemetry(): boolean {
+        return this.credentialsType === 'iam' || (this.loginType === 'builderId' && this.optOutPreference === 'OPTOUT')
+    }
+
+    private invokeSendTelemetryEvent(event: any) {
+        const request: SendTelemetryEventRequest = {
+            telemetryEvent: {
+                userTriggerDecisionEvent: event,
+            },
+        }
+        if (this.userContext !== undefined) {
+            request.userContext = this.userContext
+        }
+        if (this.optOutPreference !== undefined) {
+            request.optOutPreference = this.optOutPreference
+        }
+        this.sendTelemetryEvent(request)
+    }
+
+    public emitUserTriggerDecision(session: CodeWhispererSession, timeSinceLastUserModification?: number) {
+        if (this.isLoginInvalidForTelemetry()) {
+            return
+        }
```

Comment:
I think this condition should be moved to `invokeSendTelemetryEvent`, as it is needed for other events

Reply:
I can do this in a follow up PR, as I added test cases basis if invokeSendTelemetryEvent should be caled or not